### ISSUE TITLE
DERT value NA

### ISSUE THREAD

**Issue by sinkRswim** [bachmannpatrick/CLVTools#163]
Hi Everyone, I have three year transactional data for customers and I am using CLVTools package (which BTW is awesome) to calculate CLV. I have included a screenshot of sample output. For a large number of customers I get DERT to be NA (and consequently CLV to be NA). I use the discount factor of 7%. Any initial ideas what I might be doing wrong? Thanks in advance!

Id | period.first | period.last | period.length | PAlive | CET | DERT | predicted.mean.spending | predicted.CLV
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | 3/12/2018 | 4/1/2021 | 159.57 | 1 | 37.45 | 26.69 | 5.29 | 141.18
2 | 3/12/2018 | 4/1/2021 | 159.57 | 1 | 25.06 | 17.86 | 8.89 | 158.76
3 | 3/12/2018 | 4/1/2021 | 159.57 | 0.994163 | 14.15 | NA | 4.85 | NA
4 | 3/12/2018 | 4/1/2021 | 159.57 | 0.954839 | 4.74 | NA | 4.80 | NA

**Comment by pschil** [comment]
This is happens mostly because the estimated parameters, likely beta, are very large. This then leads to issues with numerical stability when calculating the hypergeometric 1F1 function required for DERT.

Can you give any more background in what kind of setting you are using our package?
Is it possible that you fit the model only on customers which were recently active (ie customers that you would consider "alive") or that your customer base is extremely loyal?

**Comment by sinkRswim** [comment]
Thank you Patrick. I am fitting a basic model. 

clv.train <- clvdata(train_data,
                            date.format="mdy",
                            time.unit = "week",
                            estimation.split = 80,
                            name.id = "panelist",
                            name.date = "Date",
                            name.price = TTL_US)

est.pnbd <- pnbd(clv.data = clv.train)

I get - 
  r    alpha        s     beta  
  1.021   19.069    1.325  209.411  

KKT1: TRUE 
KKT2: TRUE 

This model is for TTL US across all channels - and it runs fine. But when I run the model for a subset of this data for "grocery" channel or "club" channel, I run into the DERT issues. One option I have is fit the model for TTL US and use the above parameters to predict CLV for other channels.  Thoughts?

**Comment by pschil** [comment]
> TTL US

I am unsure how relevant the models here are for an engineering business. They are generally used for low-involvement transactions such as retail shopping. If your CLV is mainly driven by spending or involves lengthy negotiations, I believe modelling transactions might not be the primary concern. From your data it looks however ok (ie CET vs mean spending)

> But when I run the model for a subset of this data for "grocery" channel or "club" channel, I run into the DERT issues. 

I assume by "run the model" you refer to both, fit and predict. Some of your groups might be much more loyal or have only very recent customers than other groups and you get a high beta for this groups only.


> One option I have is fit the model for TTL US and use the above parameters to predict CLV for other channels

If you run the model separately for each channel, it will help you to better inspect the groups through the average transaction and dropout rates. The fitted parameters are for (shape-rate parametrized) gamma distributions and can be interpreted as r/alpha = mean transaction rate and s/beta = mean dropout rate. In your case, customers on average transact ca every 18.5 weeks, and have an average dropout rate of 0.0063 (on average live for ca 158 weeks).

If you run the model at once over all groups, the parameters it finds are something like the "weighted-average" over the parameters of each separate group.
On our own applications, we have found that splitting into groups does not help with predictive performance as measured by MAE/RMSE between CET and actual.num.trans. This might be different for you.

Alternatively, the model can be run on the whole data set and group membership can be added as covariates. This is probably the most straight-forward solution for you. See `?SetStaticCovariates` for details on how to. 
Conceptually, this is similar to splitting in groups. Through the covariate, each group receives its own alpha and beta parameter but the s and r parameters are the same across groups. 
We have found the effects on the predictive performance of including such covariates often as marginal but would be very interested to hear if it makes a difference to you.
